### PR TITLE
config: make query timeouts live update-able

### LINKED ISSUE #12232: Let's make timeouts live update-able
Can we make
```
cas_contention_timeout_in_ms
read_request_timeout_in_ms
write_request_timeout_in_ms
```
(or any other query timeouts that make sense to join the bandwagon)

live update-able?
Pretty please?

thank you in advance :-)
L

### PR DESCRIPTION
**Pull request by tchaikov**
in this change, following query timeouts config options are marked live update-able:

- range_request_timeout_in_ms
- read_request_timeout_in_ms
- counter_write_request_timeout_in_ms
- cas_contention_timeout_in_ms
- truncate_request_timeout_in_ms
- write_request_timeout_in_ms
- request_timeout_in_ms

as per https://github.com/scylladb/scylladb/issues/10172,

> Many users would like to set the driver timers based on server timers.
> For example: expire a read timeout before or after the server read time
> out.

with this change, we are able to set the timeouts on the fly. these timeout options specify how long coordinator waits for the completion of different kinds of operations. but these options are cached by the servers consuming them, so in this series, helpers are added to update the cached values when the options gets modified. also, since the observers are not copyable, sharded_parameter is used to initialize the config when creating these sharded servers.

Fixes #12232
Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

### CODE REVIEW

**Comment by avikivity** [comment]
This may not be enough. If some code copied the value of one of these config items (haven't checked, but I remember that we do), the copied value will not be refreshed. 

We have utils::updateable_value to make it work.

**Comment by tchaikov** [comment]
@avikivity thank you for your review Avi! yeah, if we cache these options locally, this would fail to update those cached versions. i will take a closer look to audit all the places referencing these options and make the necessary changes. in the meanwhile, the change is marked a draft.

**Comment by bhalevy** [comment]
The task manager ttl is probably the simplest example of using the updatable value.

See:

https://github.com/scylladb/scylladb/blob/5a07641b955c57c36cad0106127cfef96d0ee676/db/config.cc#L909

https://github.com/scylladb/scylladb/blob/5a07641b955c57c36cad0106127cfef96d0ee676/db/config.cc#L909

https://github.com/scylladb/scylladb/blob/5a07641b955c57c36cad0106127cfef96d0ee676/tasks/task_manager.hh#L51-L53

https://github.com/scylladb/scylladb/blob/5a07641b955c57c36cad0106127cfef96d0ee676/tasks/task_manager.cc#L282-L284

https://github.com/scylladb/scylladb/blob/5a07641b955c57c36cad0106127cfef96d0ee676/api/task_manager_test.cc#L98-L100

**Comment by bhalevy** [comment]
Oh, and more crucial part of using updatable_value with a sharded service is the following:

https://github.com/scylladb/scylladb/blob/5a07641b955c57c36cad0106127cfef96d0ee676/main.cc#L1017-L1022

File: db/config.cc
Comment (xemul) [inline db/config.cc:569]: Why is this one not included?

File: db/config.cc
Comment (xemul) [inline db/config.cc:573]: And this

**Comment by bhalevy** [comment]
Regarding shard safety - that's the reason I mentioned `tasks::task_manager::config` passed as a `sharded_parameter` in https://github.com/scylladb/scylladb/pull/12531#issuecomment-1384401454

Let's see if `timeout_config` can keep `utils::updateable_value` members in a similar way and passed around as a `sharded_parameter`

**Comment by tchaikov** [comment]
hi folks, thanks for your insights. i was wrapped up in some other issues and later on PTO, will be back to this one in this week.

**Comment by tchaikov** [comment]
changelog

- mark two more timeout options with `liveness::LiveUpdate`
- drop unused `timeout_config()` methods
- remove unused forward declaration
- pass const reference to thrift's factory
- create redis, thrift and cql_server with sharded_parameter
- add helpers to keep timeout_config updated
- use designated initializer when appropriate

**Comment by tchaikov** [comment]
@xemul @bhalevy hi Pavel and Benny, could you take another look?

**Comment by nyh** [comment]
> This may not be enough. If some code copied the value of one of these config items (haven't checked, but I remember that we do), the copied value will not be refreshed.
> 
> We have utils::updateable_value to make it work.

Right. `alternator_timeout_in_ms` (which you missed, by the way, deliberately?) unfortunately has in `alternator/controller.cc` the following call once, during the boot of the server:

```c++
        executor::set_default_timeout(std::chrono::milliseconds(_config.alternator_timeout_in_ms()));
```

So any later changes to config.alternator_timeout_in_ms will be ignored :-( To support this we'll need to change executor to store the updateable_value, or something like that...

**Comment by tchaikov** [comment]
changelog

- added a commit to make `alternator_timeout_in_ms` live-updateable.

File: timeout_config.cc
Comment (xemul) [inline timeout_config.cc:20]: Somewhere on stop there should be call to serializeable_action::join() to make sure no outstanding updates are scheduler, but I don't see any

File: timeout_config.hh
Comment (xemul) [inline timeout_config.hh:25]: Isn't it fragile to leave these numbers as-is? E.g. previous version lost alternator copy of some of those values. Why not patch _all_ the above fields to be `updateable_value<duration>`s and patch the users/callers/passers/etc. accordingly? The updateable_value can be easily subscribed on db::config::named_value change

Comment (xemul) [reply timeout_config.hh:25]: Yes. Look at, e.g., `struct restrictions_config`. It has two "integers" onboard, booth are live-updateable via db::config. It's exactly what this PR is trying to achieve with timeout_config and my point is that converting all its inhabitants to be `updateable_value<duration>` is the safest way to go

Comment (tchaikov) [reply timeout_config.hh:25]: i see. thanks! makes sense to me.

**Comment by tchaikov** [comment]
marking as a draft. will update according to Pavel's suggestion.

File: redis/server.cc
Comment (Deexie) [inline redis/server.cc:39]: Use `_config` here as `config` is already moved

Comment (Deexie) [reply redis/server.cc:39]: Same for `_total_redis_db_count`

**Comment by bhalevy** [comment]
@xemul please re-review / merge

**Comment by nyh** [comment]
Please update the cover letter to also mention the Alternator timeout, or if you find additional ones you forgot to list.

File: transport/server.hh
Comment (xemul) [inline transport/server.hh:215]: nit: That's not `cql_server::timeout_config`, but `cql_server::connection::timeout_config`. The former is in fact used and you patch it to be const with https://github.com/scylladb/scylladb/pull/12531/commits/a1872332fe77475bdd06f897339ac59b4114af07

File: db/config.cc
Comment (xemul) [inline db/config.cc:566]: nit: s/anther/another/ in commit message

File: db/config.cc
Comment (xemul) [inline db/config.cc:883]: Do I get this right, that this particular patch has nothing to do with `timeout_config` and it _just_ makes the `alternator_timeout_in_ms` live-updateable with all the necessary changed in alternator code to support the update?

File: timeout_config.hh
Comment (xemul) [inline timeout_config.hh:23]: Unused

File: timeout_config.hh
Comment (xemul) [inline timeout_config.hh:20]: This struct is left because things like client_state (and thrift_handler, but nobody cares about thrift nowadays) carry _it_ onboard, not the updated_timeout_config. So the whole timeout-config-thing becomes life-updateable in a sense that _new_ connections obey that, while old/existing ones continue working with older timeouts.

I think it deserves at least a comment somewhere in timeout_config.hh, because it's not immediately obvious that this is how things should work.

File: timeout_config.hh
Comment (xemul) [inline timeout_config.hh:59]: Wow. This caught my eye. As far as I know the `updateable_value<T>` has `operator()` on board, so if we do `updatable_timeout_config.read_timeout_in_ms()` it will work by returning the current value as plain `duration_t`. Thus, this helper (and the `write_timeout()` below too) are excessive, as any code that needs to get just the timeout immediate value -- it's `redis_options::get_read_timeout()` for example -- may just call `::operator()` on it and that's it.

File: timeout_config.hh
Comment (xemul) [inline timeout_config.hh:48]: This constructor is unused

File: service/client_state.hh
Comment (xemul) [inline service/client_state.hh:164]: nit: I appreciate changes like that, but this particular one is alien to this patch

File: timeout_config.hh
Comment (tchaikov) [reply timeout_config.hh:23]: removed. it was an leftover from previous refactory.

File: timeout_config.hh
Comment (tchaikov) [reply timeout_config.hh:20]: good point. i made some notes in the comment. but this should be reflected in the code as well.

**Review by xemul — CHANGES_REQUESTED** [review]
Thanks for pushing this forward! I've some cosmetic comments, other than that I think it's ready to get merged

File: timeout_config.hh
Comment (xemul) [reply timeout_config.hh:20]: Thanks!

File: alternator/executor.hh
Comment (nyh) [inline alternator/executor.hh:212]: I think, or rather hope, that this change isn't needed. Why did you need to "disconnect"? Is it to solve some destruction-order issues? Shouldn't the config be the LAST thing to destroy, after Alternator and all other front-ends have finished their own destruction?
This code isn't terrible (just one line of code) but it ugly and the fact it's needed might hint we have other bugs we just haven't found yet.

File: db/config.cc
Comment (nyh) [reply db/config.cc:883]: If I understand correctly, "timeout_config" was some sort of class that covered some random selection of timeout-related parameters in the configuration, and alternator_timeout_in_ms was never part of it.
To be honest, I don't understand why a timeout_config class was ever needed, it sounds to me that every piece of code needs to request one specific timeout, and there is never a need to get the entire structure. But maybe I'm missing something?

File: timeout_config.hh
Comment (tchaikov) [reply timeout_config.hh:34]: okay

File: timeout_config.cc
Comment (tchaikov) [reply timeout_config.cc:48]: sure. the less code the merrier. i missed this in previous refactory.

File: db/config.cc
Comment (nyh) [reply db/config.cc:883]: We already have a "set of configuration parameters that Scylla supports", I don't see any reason why we need special subsets for it, like the "list of timeouts", the "list of ports", and so on. I think it just complicates everything. I think what we *should* have done is for each server to have a pointer to the configuration of the current shard - and the server could use this configuration to get anything from it. I think the current design is a mistake (but of course, no need to fix it in this PR, this is not a new issue).

In particular, if we don't want Scylla to become one huge spaghetti code, there is no reason to bundle CQL features, Redis features, and Alternator features together just because they sound superficially similar. Yes, CQL, Redis and Alternator all have some sort of timeout options - but these are *different* options, with different meanings, different defaults, etc., so bundling all of them together just makes the code harder to decipher.

Comment (xemul) [reply db/config.cc:883]: Yes, you're right. And we have per-service config to collect the needed parameters from db::config and pull them down the service. I guess the timeout_config was introduce to generalize this between transport, redis and thrift.

Comment (tchaikov) [reply db/config.cc:883]: > We already have a "set of configuration parameters that Scylla supports", I don't see any reason why we need special subsets for it, like the "list of timeouts", the "list of ports", and so on. I think it just complicates everything. I think what we _should_ have done is for each server to have a pointer to the configuration of the current shard - and the server could use this configuration to get anything from it. I think the current design is a mistake (but of course, no need to fix it in this PR, this is not a new issue).
> 
> In particular, if we don't want Scylla to become one huge spaghetti code, there is no reason to bundle CQL features, Redis features, and Alternator features together just because they sound superficially similar. Yes, CQL, Redis and Alternator all have some sort of timeout options - but these are _different_ options, with different meanings, different defaults, etc., so bundling all of them together just makes the code harder to decipher.

@nyh i think we can implement in a follow-up PR, which 

1. pushes the `updateable_timeout_config` down to their owner, namely, the config bundles in redis, thrift and CQL servers.
2. replace the bundles with a reference to `db::config`.
 
if we are all for this change.

File: timeout_config.hh
Comment (tchaikov) [reply timeout_config.hh:48]: dropped.

File: service/client_state.hh
Comment (tchaikov) [reply service/client_state.hh:164]: i will piggy back it as a dedicated commit in this patchset.

File: redis/controller.cc
Comment (tchaikov) [reply redis/controller.cc:52]: > Nitpick: Since we already have the name **updateable** value, no **updated** value, I suggest that we should use the prefix updatable_ everywhere, not updated_.

the same suggestion was proposed by @xemul , see https://github.com/scylladb/scylladb/pull/12531#discussion_r1144522389 . and i still think you guys suggestions make sense.

yeah, the explanation helps to convince me that updateable is better. thanks. the naming is difficult.

File: transport/server.hh
Comment (tchaikov) [reply transport/server.hh:272]: yeah, `client_state` 's ctor "consumes" the `::timeout_config` returned by this function as an xvalue, and it is eligible to have copy elision, so we should be fine from the performance perspective.

File: db/config.cc
Comment (tchaikov) [reply db/config.cc:883]: > I see. I would then suggest submitting this patch as separate PR

okay. but can we continue the discussion related to  `alternator_timeout_in_ms` , as @nyh already started review it and post some comments on it.

File: alternator/executor.cc
Comment (tchaikov) [reply alternator/executor.cc:1370]: > I don't understand the commit message's discussion on this default, or why you think it's related to testing only - this is the real default in the code unless the yaml overrides this default...


probably i missed something. but i think, despite that  `executor::default_timeout()` is a static function, it is always called *after* the executor is up and running. as we start the controller server in `main()`. so, its default value is always overridden in the constructor no matter if the yaml conffile has a different value or not.

> Also please change the comment above to be up-to-date with the changed names (and also the changed behavior, if there is one).

ahh, sure. updated in the latest revision.

File: alternator/executor.hh
Comment (tchaikov) [reply alternator/executor.hh:180]: sure, makes sense.

File: alternator/executor.hh
Comment (tchaikov) [reply alternator/executor.hh:212]: > I think, or rather hope, that this change isn't needed. Why did you need to "disconnect"? Is it to solve some destruction-order issues?

i added it just for the sake of symmetry. as i thought, if the constructor connects `updateable_value` to the one in the conffile, we should perform the opposite operation when tearing it down, another option is do this in the dtor.

> Shouldn't the config be the LAST thing to destroy, after Alternator and all other front-ends have finished their own destruction? This code isn't terrible (just one line of code) but it ugly and the fact it's needed might hint we have other bugs we just haven't found yet.

i agree that config is the last one to destroy. so we don't need to worry about dangling reference. 
if you still think we should drop this. i am happy to drop this line.

File: test/lib/alternator_test_env.cc
Comment (tchaikov) [reply test/lib/alternator_test_env.cc:27]: ahh, true! changed accordingly.

File: timeout_config.cc
Comment (tchaikov) [reply timeout_config.cc]: this does not matter anymore. since i just use the updateable_value instead of the observer + action machinery for updating a local duration value.

**Comment by tchaikov** [comment]
changelog

- correct the commit message: s/cql_server::timeout_config/cql_server::connection::timeout_config/
- correct the typo in commit message: s/anther/another/
- s/updated_timeout_config/updateable_timeout_config/
- drop the unused `from_updated()` 
- drop the unused `make_timeout_config()`
- drop the unused `updated_timeout_config(duration<>)` constructor
- drop the `read_timeout()` and `write_timeout()` accessors, let the caller to convert a plain number a duration.
- add note on the connection between `timeout_config` and `updateable_timeout_config`
- extract the changes which split the long lines in ctor into multiple shorter lines to separate commits
- update the comment above `s_default_timeout_in_ms` 's definition to reflect the name change.
- s/timeout_in_ms/default_timeout_in_ms/ in the ctor of alternator's `executor`
- use the much shorter `std::chrono::milliseconds ms = 10s;` in test. thanks!

not changed:

- the ugly line in `executor::stop()` is not removed. i will remove it if @nyh insists.
- extract the `alternator_timeout_in_ms` related changes into another PR: i will do so once the comments on them are resolved.
- remove `timeout_config` and even the config bundles used by different servers, and just use a reference to `db::config`, we should do this in a separate PR after we have a consensus.

**Review by xemul — APPROVED** [review]

**Comment by tchaikov** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/262/

#12629

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help queue this changeset?

**Comment by xemul** [comment]
@tchaikov , merge conflict, please rebase

**Comment by tchaikov** [comment]
> @tchaikov , merge conflict, please rebase

@xemul, hi Pavel, thanks for queuing this patchset. i think it conflicts with #13319, which was included in next before you tried to merge mine. probably i have to wait until it gets merged, and rebase onto master.

**Comment by xemul** [comment]
@tchaikov , it looks like conflicting PR is already in master. Please, rebase this one one more time :)

**Pull request merged**
